### PR TITLE
Add stateful for loop primitives

### PR DESCRIPTION
**Pull request by sharadmv**
Adds a `get/swap/addupdate` primitive, along with impl, abstract_eval and jvp rules.

Tracking issue:  #10982

### CODE REVIEW

**Comment by sharadmv** [comment]
@mattjj thoughts on how to test the get/set transpose rules?

File: jax/_src/lax/control_flow/for_loop.py [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+from jax._src import ad_util
+from jax._src import pretty_printer as pp
+
+
+## Helpful type aliases
+Ref = Array = Any
+
+## State effect
+
+State = object()
```
Comment (mattjj) [inline jax/_src/lax/control_flow/for_loop.py:30]:
```suggestion
class StateEffect: pass
State = StateEffect
```

Leads to nicer pdb printing if/when necessary!

File: jax/_src/lax/control_flow/for_loop.py [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+## get/swap/addupdate implementations
+
+# `get` reads a value reading a value from a `Ref` type, a.k.a.:
+# a <- x[] -- a <=> get_p.bind(x, idx=())
```
Comment (mattjj) [inline jax/_src/lax/control_flow/for_loop.py:36]: I'm not sure about these comments; the the thing to the left of `<=>` isn't jaxpr syntax, and it's not Python syntax, so what's it meant to be? And then what's `<=>` meant to mean exactly? (And is the `-- a` a typo?)

My guess is it's meant to be jaxpr syntax, but in that case I'd revise it to be more precise, like

```
# Staging out `a = get_p.bind(x, idx=())` where the aval of `x` is
# `ShapedArrayRef((3,), np.dtype('float32'))` leads to a jaxpr eqn printed like
#   b:f32[] <- x[]
```

WDYT?

Comment (mattjj) [reply jax/_src/lax/control_flow/for_loop.py:36]: If the comment is about jaxpr pretty-printing, maybe it should be moved down to the pretty-printing section?

If it's instead about the semantics of the get operation and maybe its idx parameter, maybe it should be defined without referring to pseudo-syntax (i.e. the `a <- x[]` stuff) and instead using words?

**Review by mattjj — APPROVED** [review]
Woo! Nice tests.

File: jax/_src/lax/control_flow/for_loop.py (hunk 2 above)
Comment (sharadmv) [reply jax/_src/lax/control_flow/for_loop.py:36]: Yep, good idea.

**Comment by sharadmv** [comment]
Yes looks like a typo.

**Pull request merged**